staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject_test.go

```diff
@@ -0,0 +1,213 @@
+		{
+			name: "pod with no protocol",
+			object: &corev1.Pod{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Ports: []corev1.ContainerPort{
+								{
+									ContainerPort: 101,
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: map[string][]string{"101": {"TCP"}},
+		},
+		{
+			name: "pod with same-port,different-protocol",
+			object: &corev1.Pod{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Ports: []corev1.ContainerPort{
+								{
+									ContainerPort: 101,
+									Protocol:      "TCP",
+								},
+								{
+									ContainerPort: 101,
+									Protocol:      "UDP",
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: map[string][]string{"101": {"TCP", "UDP"}},
+		},
+		{
+			name: "service with TCP protocol",
+			object: &corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Port:     101,
+							Protocol: "TCP",
+						},
+					},
+				},
+			},
+			expected: map[string][]string{"101": {"TCP"}},
+		},
+		// No protocol for service port--default to TCP
+		{
+			name: "service with no protocol",
+			object: &corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Port: 101,
+						},
+					},
+				},
+			},
+			expected: map[string][]string{"101": {"TCP"}},
+		},
+		{
+			name: "replication with TCP protocol",
+			object: &corev1.ReplicationController{
+				Spec: corev1.ReplicationControllerSpec{
+					Template: &corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Ports: []corev1.ContainerPort{
+										{
+											ContainerPort: 101,
+											Protocol:      "TCP",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: map[string][]string{"101": {"TCP"}},
+		},
+		{
+			name: "deployment with TCP protocol",
+			object: &extensionsv1beta1.Deployment{
+				Spec: extensionsv1beta1.DeploymentSpec{
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Ports: []corev1.ContainerPort{
+										{
+											ContainerPort: 101,
+											Protocol:      "TCP",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: map[string][]string{"101": {"TCP"}},
+		},
+		{
+			name: "replicaset with TCP protocol",
+			object: &extensionsv1beta1.ReplicaSet{
+				Spec: extensionsv1beta1.ReplicaSetSpec{
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Ports: []corev1.ContainerPort{
+										{
+											ContainerPort: 101,
+											Protocol:      "TCP",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: map[string][]string{"101": {"TCP"}},
+		},
+		{
+			name:      "unsupported object",
```

Comment:
Does `portsforobject` also return `expectErr` true?

Reply:
yes

This is the code for portsforobject

```go
func portsForObject(object runtime.Object) ([]string, error) {
	switch t := object.(type) {
	case *corev1.ReplicationController:
		return getPorts(t.Spec.Template.Spec), nil

	case *corev1.Pod:
		return getPorts(t.Spec), nil

	case *corev1.Service:
		return getServicePorts(t.Spec), nil

	case *extensionsv1beta1.Deployment:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1.Deployment:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1beta2.Deployment:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1beta1.Deployment:
		return getPorts(t.Spec.Template.Spec), nil

	case *extensionsv1beta1.ReplicaSet:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1.ReplicaSet:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1beta2.ReplicaSet:
		return getPorts(t.Spec.Template.Spec), nil
	default:
		return nil, fmt.Errorf("cannot extract ports from %T", object)
	}
}
```